parser-typechecker/src/Unison/Codebase/Branch.hs

```diff
@@ -211,80 +222,161 @@ branch0 terms types children edits =
-deriveDeepPaths :: Branch0 m -> Branch0 m
+deriveDeepPaths :: forall m. Branch0 m -> Branch0 m
 deriveDeepPaths branch =
-  branch {deepPaths = makeDeepPaths (nonEmptyChildren branch)}
+  branch {deepPaths = makeDeepPaths branch}
   where
-    makeDeepPaths :: Map NameSegment (Branch m) -> Set Path
-    makeDeepPaths children =
-      Set.mapMonotonic Path.singleton (Map.keysSet children) <> ifoldMap go children
+    makeDeepPaths :: Branch0 m -> Set Path
+    makeDeepPaths branch = flip State.evalState mempty $ go [(mempty, branch)] mempty
       where
-        go :: NameSegment -> Branch m -> Set Path
-        go n b =
-          Set.map (Path.cons n) (deepPaths $ head b)
+        go :: [([NameSegment], Branch0 m)] -> Set Path -> DeepState m (Set Path)
+        go [] acc = pure acc
+        go ((reversePrefix, b0) : work) acc = do
+          let paths :: Set Path
+              paths =
+                if isEmpty0 b0
+                  then Set.empty
+                  else (Set.singleton . Path . Seq.fromList . reverse) reversePrefix
+          children <- deepChildrenHelper reversePrefix b0
+          go (work <> children) (paths <> acc)
 
 -- | Derive the 'deepEdits' field of a branch.
-deriveDeepEdits :: Branch0 m -> Branch0 m
+deriveDeepEdits :: forall m. Branch0 m -> Branch0 m
 deriveDeepEdits branch =
-  branch {deepEdits = makeDeepEdits (_edits branch) (nonEmptyChildren branch)}
+  branch {deepEdits = makeDeepEdits branch}
   where
-    makeDeepEdits :: Map NameSegment (EditHash, m Patch) -> Map NameSegment (Branch m) -> Map Name EditHash
-    makeDeepEdits edits children =
-      Map.mapKeysMonotonic Name.fromSegment (Map.map fst edits) <> ifoldMap go children
+    makeDeepEdits :: Branch0 m -> Map Name EditHash
+    makeDeepEdits branch = flip State.evalState mempty $ go [(mempty, branch)] mempty
       where
-        go :: NameSegment -> Branch m -> Map Name EditHash
-        go n b =
-          Map.mapKeys (Name.cons n) (deepEdits $ head b)
+        go :: [([NameSegment], Branch0 m)] -> Map Name EditHash -> DeepState m (Map Name EditHash)
+        go [] acc = pure acc
+        go ((reversePrefix, b0) : work) acc =
+          let edits :: Map Name EditHash
+              edits =
+                Map.mapKeysMonotonic
+                  (Name.fromReverseSegments . (NonEmpty.:| reversePrefix))
+                  (fst <$> _edits b0)
+           in do
+                children <- deepChildrenHelper reversePrefix b0
+                go (work <> children) (edits <> acc)
+
+type DeepState m = State (Set (NamespaceHash m), Sum Int)
+
+-- | Helper for knowing whether to descend into a child branch or not.
+-- Accepts child namespaces with previously unseen hashes, and any nested under fewer than 2 `lib` segments.
+deepChildrenHelper :: [NameSegment] -> Branch0 m -> DeepState m [([NameSegment], Branch0 m)]
+deepChildrenHelper reversePrefix b0 = do
+  case reversePrefix of
+    "lib" : _ -> modifying _2 (+ 1)
```

Comment:
Since we're in the state monad, we'll continue to accumulate this number as we go; I don't think I understand exactly how your recursive `go` works, but if it follows either of BFS or DFS I think this strategy is incorrect, since one branch of the tree affects the state for ALL branches traversed after it.

Maybe using a reader monad with `local` and explicitly recursing rather than tail-recursion would be a good fix?

Reply:
I managed to find a transcript that demonstrates the issue, notice that if we've encountered at least two libs, then AFTER that we find two identical namespaces, the second will be ignored even though it's not inside a `lib`;

So in the final names call I expect to see both `b.b.b.x` and `c.c.c.x`, but only `b.b.b.x` is shown.

    ```unison
    lib.x = 1
    a.lib.x = 2
    b.b.b.x = 5
    c.c.c.x = 5
    ```

    ```ucm

      I found and typechecked these definitions in scratch.u. If you
      do an `add` or `update`, here's how your codebase would
      change:
      
        ⍟ These new definitions are ok to `add`:
        
          a.lib.x : ##Nat
          b.b.b.x : ##Nat
          c.c.c.x : ##Nat
          lib.x   : ##Nat

    ```
    ```ucm
    .> add

      ⍟ I've added these definitions:
      
        a.lib.x : ##Nat
        b.b.b.x : ##Nat
        c.c.c.x : ##Nat
        lib.x   : ##Nat

    .> names x

      Term
      Hash:   #k09e57l61r
      Names:  b.b.b.x
      
      Tip: Use `names.global` to see more results.

    ```

Reply:
Yeah this is the bug I converted back to Draft for 😞

Reply:
Let me know if you'd like a hand fixing it, I have some ideas 😄

Reply:
Ok, this is fixed in the latest:

````
```unison
lib.x = 1
a.lib.x = 2
b.b.b.x = 5
c.c.c.x = 5
```

```ucm

  I found and typechecked these definitions in scratch.u. If you
  do an `add` or `update`, here's how your codebase would
  change:
  
    ⍟ These new definitions are ok to `add`:
    
      a.lib.x : ##Nat
      b.b.b.x : ##Nat
      c.c.c.x : ##Nat
      lib.x   : ##Nat

```
```ucm
.> add

  ⍟ I've added these definitions:
  
    a.lib.x : ##Nat
    b.b.b.x : ##Nat
    c.c.c.x : ##Nat
    lib.x   : ##Nat

.> names x

  Term
  Hash:   #k09e57l61r
  Names:  b.b.b.x c.c.c.x
  
  Tip: Use `names.global` to see more results.

```
````